Cfstore 3.0.0

shroffk opened the pull request:
Preparation branch for testing a PR to the main recsync project

mskinner5278 committed:
Enforced standard python formatting

mskinner5278 committed:
Improvements to cf-store as defined in issue #11

mskinner5278 committed:
Unit tests for cf-store, including mock channelfinder client. Tests currently focused on IOC disconnections.

mskinner5278 committed:
Added new tests for cf-store, including those for channelfinder client disconnections.

mskinner5278 committed:
Fixed cases where commit attempted to make no changes and would poll indefinitely. Fixed issue where recceiver was ignoring keepalive response and causing a timeout after 30 seconds. Added more cases to commit order tests.

mskinner5278 committed:
Added locking to cf-store deferred. Added script for outputting channelfinder service data to a file, same format as cf-store data file print. Added script for creating 1000~ identical iocs using the recsync demo st.cmd.

mskinner5278 committed:
'channelcount' is now used to clean ioc dictionary, writes to error log if it becomes negative. Glassfish server errors are now caught as requests.exceptions.HTTPError instead of StandardError. Mock_client now also uses HTTPError. Mock client uses IPV4Address as its source, rather than an arbitrary object. Recast no longer pauses accepting input, also removed re-initiating listening. test_cfstore cleaned up threading and test commits.

mskinner5278 committed:
Added cleaning methods to mock client, added tests for clean_service to unit tests. Changed format of IOCID to match possible new system. Updated owner information. Clean service now uses the update method instead of the set methods. Note: This overwrites ownership in some locations.

mskinner5278 committed:
Added readme to describe and monitor the status of manual tests. Fixed local path in test_mock_iocs and decreased the number of iocs spawned by default. Cleaned up the dictionary logic in the commit method. Added config control to dict_to_file.

mskinner5278 committed:
Added new test IOC "st_test.cmd" with 1000 channels. Added script for adding active channels for manual tests. Added readme to describe manual test process. Added debug logging for cleaning stages.

mskinner5278 committed:
Added cf-store section to conf file, commented out debug file production

mskinner5278 committed:
Now handling all instances of RequestException. Cleaned up dictionary logic with collections.defaultdict.

mskinner5278 committed:
Removed useless code and explicit log level checking.

shroffk committed:
cfstore updated to reading info tags and use new cf client

shroffk commented:
@dylan171 @mdavidsaver 
Reviewed most of the changes...still not sure about a few things due to my lack of expertise...

My last commit added a few major changes.
1. I replaced certain individual CF calls with a single one, this should result in hundreds of fewer n/w calls depending on the number of channels.
2. IOC developers can add info tags with the syntax 
   info (properties, "propertyName=propertyValue,...")
   e.g.
   `info(properties, "device=$(M)")`

mdavidsaver commented on server/recceiver/cfstore.py:
```diff
@@ -69,15 +70,49 @@ def commit(self, transaction_record):
+                        properties.append({u'name': p[0], u'owner': owner, u'value': p[1]})
+                    pvInfo[rid]["infoProperties"] = properties
+                else:
+                    _log.exception("could not find the associated record for properties")
```
`.exception()` is only useful when called when there is an active exception (eg. within a `catch:` block).  Otherwise just use `.error()`.

mdavidsaver commented on server/recceiver/cfstore.py:
```diff
@@ -69,15 +70,49 @@ def commit(self, transaction_record):
+        pvNames = [info["pvName"] for rid, (info) in pvInfo.iteritems()]
+        
+        delrec = list(TR.delrec)
+        _log.info(delrec)
```
Should have some prefix string

mdavidsaver commented on server/recceiver/cfstore.py:
```diff
@@ -86,17 +121,21 @@ def __commit__(self, TR):
                 self.channel_dict[pv].remove(iocid)
-                self.iocs[iocid]["channelcount"] -= 1
+                if iocid in self.iocs:
+                    self.iocs[iocid]["channelcount"] -= 1
```
Is the added test here from paranoia, or due to an issue encountered?  I'm a bit nervous about the manual ref. counter, and would like to be clear if there is any issue discovered.

shroffk replied:
I am not completely comfortable with this either. I did have an issue when doing my tests (restarting the same ioc on the same machine multiple times) which resulted in this check...
The main goal of my changes is to minimize the network calls and to allow the parsing of info tags in the iocs

mdavidsaver commented on server/recceiver/cfstore.py:
```diff
@@ -134,133 +173,101 @@ def dict_to_file(dict, iocs, conf):
+                     {u'name': 'pvStatus', u'owner': owner, u'value': "Active"},
+                     {u'name': 'time', u'owner': owner, u'value': iocTime}]
+        infoProperties = [info["infoProperties"] for rid, (info) in pvInfo.iteritems() if info["pvName"] == pv and "infoProperties" in info ]
+        _log.debug(infoProperties)
```
same here.

mdavidsaver reviewed (approved):
Overall looks good.  A few minor points about logging, please fix if you have time, otherwise I will eventually.

shroffk committed:
#15 addressing mdavidsavers comments

dxmaxwell committed:
Exception thrown by log message

In my testing this line threw an exception: exceptions.TypeError: cannot concatenate 'str' and 'list' objects

shroffk committed:
Merge pull request #1 from dylan171/patch-3

Exception thrown by log message

shroffk commented:
@dylan171  @mdavidsaver  can we merge this by the end of this week?

dxmaxwell commented:
Just deployed the latest cfstore for testing and ran into this error:

2016-11-03 15:22:16-0400 [CastReceiver,0,10.41.200.35] DEBUG:urllib3.connectionpool "GET /channelfinder//resources/channels?~name=_FE_LEBT%3ARFC1_D0987%3Areconnected%7C_FE_LEBT%3ARFC2_D0987%3Areconnected%7C_FE_LEBT%3ARFC2_D0987%3AfwdAmplDev%7CFE_L
EBT%3ARFC1_D0987%3AHLIM_RD_CAV%7Ciocllrf%3AAUTOSAVE%3ASR_3_State%7CFE_LEBT%3ARFC2_D0987%3AILK_RSTS_AFHI%7C_FE_LEBT%3ARFC1_D0987%3Aset_pulserOn%7CFE_LEBT%3ARFC1_D0987%3APOS_RSET_TRIG.......
2016-11-03 15:22:16-0400 [CastReceiver,0,10.41.200.35] DEBUG:recceiver.cfstore error: 414 Client Error: Request-URI Too Long

shroffk commented:
I am just workin on this and had hoped to sneak a fix in before anyone noticed 
:(
With the new cfstore, while I have reduced the number of n/w calls by combining each individula channle calls into a single one....this can significantly reduce the number of times you have to query CF... however if the ioc has many pv's the above error was likely. I apologize for the shoddiness ...

shroffk committed:
Breaking up the URL when it exceeds 600 chars

dxmaxwell committed:
Fix failing unit test

shroffk committed:
Merge pull request #2 from dylan171/cfstore-3.0.0

Fix failing unit test

dxmaxwell commented:
@mdavidsaver, I discuss the ownership of the RecSync repository with @shroffk and we agreed that the ChannelFinder organization would be the best fit right now.

Please merge this PR and then transfer the repository to ChannelFinder. Thanks.

mdavidsaver commented:
> Please merge this PR and then transfer the repository to ChannelFinder. Thanks.

Step 1 complete.  Some problems with step 2.

```
ChannelFinder/recsync already exists and You don’t have admin rights to ChannelFinder 
```
